src/twisted/trial/test/matchers.py

```diff
@@ -21,17 +23,46 @@ class _MatchAfter(BaseMatcher[_A]):
 
     @ivar f: The function to apply.
     @ivar m: The matcher to use on the result.
+
+    @ivar _e: After trying to apply the function fails with an exception, the
+        exception that was raised.  This can later be used by
+        L{describe_mismatch}.
     """
 
     def __init__(self, f: Callable[[_A], _B], m: Matcher[_B]):
         self.f = f
         self.m = m
+        self._e: Optional[Exception] = None
 
     def _matches(self, item: _A) -> bool:
         """
         Apply the function and delegate matching on the result.
         """
-        return self.m.matches(self.f(item))
+        try:
+            transformed = self.f(item)
+        except Exception as e:
+            self._e = e
+            return False
+        else:
+            return self.m.matches(transformed)
+
+    def describe_mismatch(self, item: _A, mismatch_description: Description) -> None:
+        """
+        Describe the mismatching item or the exception that occurred while
+        pre-processing it.
+
+        @note: Since the exception reporting here depends on mutable state it
+            will only work as long as PyHamcrest calls methods in the right
+            order.  The PyHamcrest Matcher interface doesn't seem to allow
+            implementing this functionality in a more reliable way (see the
+            implementation of L{assert_that}).
+        """
+        if self._e is None:
+            super().describe_mismatch(item, mismatch_description)
+        else:
+            mismatch_description.append_text(
+                f"{self.f.__qualname__}({item!r}) raised\n    {self._e}"
```

Comment:
This should include the exception name, not just the message.

Reply:
Done.  I ended up taking the message _out_ of the assertion made by the test (and only using the exception class name) because the message could vary based on locale.